### PR TITLE
Add the option to generate HTML fragments and the body of an HTML doc…

### PR DESCRIPTION
**Pull request by rudkjobing**
Implement feature requested here: https://github.com/fzaninotto/Faker/pull/971#issuecomment-285607794

### CODE REVIEW

**Comment by rudkjobing** [comment]
@fzaninotto Hmm is master broken? Tests wont run for me after i merge master into my fork.

`Testing started at 10:18 PM ...
<?xml version="1.0" encoding="UTF-8"?><contacts>
PHP Notice:  Undefined offset: -32 in C:\SLFaker\src\Faker\Provider\Base.php on line 193
PHP Notice:  Undefined index:  in C:\SLFaker\src\Faker\Provider\Base.php on line 193
PHP Notice:  Undefined offset: -229 in C:\SLFaker\src\Faker\Provider\Base.php on line 193
PHP Notice:  Undefined index:  in C:\SLFaker\src\Faker\Provider\Base.php on line 193
PHP Fatal error:  Uncaught Exception: domainWord failed with the selected locale. Try a different locale or activate the "intl" PHP extension. in C:\SLFaker\src\Faker\Provider\Internet.php:138
Stack trace:
#0 C:\SLFaker\src\Faker\Provider\Internet.php(124): Faker\Provider\Internet->domainWord()
#1 [internal function]: Faker\Provider\Internet->domainName()
#2 C:\SLFaker\src\Faker\Generator.php(207): call_user_func_array(Array, Array)
#3 C:\SLFaker\src\Faker\Generator.php(241): Faker\Generator->format('domainName')
#4 [internal function]: Faker\Generator->callFormatWithMatches(Array)
#5 C:\SLFaker\src\Faker\Generator.php(236): preg_replace_callback('/\\{\\{\\s?(\\w+)\\s...', Array, '{{userName}}@{{...')
#6 C:\SLFaker\src\Faker\Provider\Internet.php(40): Faker\Generator->parse('{{userName}}@{{...')
#7 [internal function]: Faker\Provider\Internet->email()
#8 C:\SLFaker\src\Faker\Generator.php(207): call_user_func_array(Array, Array)
#9 C:\SLFaker\src\Faker\Generator.php(249): Faker\Generator->format('email')
#10 C:\SLFaker\test\te in C:\SLFaker\src\Faker\Provider\Internet.php on line 138
  <contact firstName="Arely" lastName="Champlin" email="
Notice: Undefined offset: -32 in C:\SLFaker\src\Faker\Provider\Base.php on line 193

Notice: Undefined index:  in C:\SLFaker\src\Faker\Provider\Base.php on line 193

Notice: Undefined offset: -229 in C:\SLFaker\src\Faker\Provider\Base.php on line 193

Notice: Undefined index:  in C:\SLFaker\src\Faker\Provider\Base.php on line 193

Fatal error: Uncaught Exception: domainWord failed with the selected locale. Try a different locale or activate the "intl" PHP extension. in C:\SLFaker\src\Faker\Provider\Internet.php:138
Stack trace:
#0 C:\SLFaker\src\Faker\Provider\Internet.php(124): Faker\Provider\Internet->domainWord()
#1 [internal function]: Faker\Provider\Internet->domainName()
#2 C:\SLFaker\src\Faker\Generator.php(207): call_user_func_array(Array, Array)
#3 C:\SLFaker\src\Faker\Generator.php(241): Faker\Generator->format('domainName')
#4 [internal function]: Faker\Generator->callFormatWithMatches(Array)
#5 C:\SLFaker\src\Faker\Generator.php(236): preg_replace_callback('/\\{\\{\\s?(\\w+)\\s...', Array, '{{userName}}@{{...')
#6 C:\SLFaker\src\Faker\Provider\Internet.php(40): Faker\Generator->parse('{{userName}}@{{...')
#7 [internal function]: Faker\Provider\Internet->email()
#8 C:\SLFaker\src\Faker\Generator.php(207): call_user_func_array(Array, Array)
#9 C:\SLFaker\src\Faker\Generator.php(249): Faker\Generator->format('email')
#10 C:\SLFaker\test\te in C:\SLFaker\src\Faker\Provider\Internet.php on line 138

Process finished with exit code 255
`

**Comment by Zhastreaus** [comment]
Sorry to be that guy, but any update on this?

File: test/Faker/Provider/HtmlLoremTest.php [hunk 1]
```diff
@@ -20,10 +20,28 @@ public function testProvider()
...
+        $dom = new \DOMDocument();
+        $error = $dom->loadHTML($node);
+        $this->assertTrue($error);
+    }
+
+    public function testRandomBodyReturnsValidHTMLString(){
+        $faker = new Generator();
+        $faker->addProvider(new HtmlLorem($faker));
+        $node = $faker->randomBody(5, 5);
+        echo $node;
```
Comment (WillGibson) [inline test/Faker/Provider/HtmlLoremTest.php]: I don't think you want the `echo $node;` line?

File: test/Faker/Provider/HtmlLoremTest.php [hunk 2]
```diff
@@ -20,10 +20,28 @@ public function testProvider()
     public function testRandomHtmlReturnsValidHTMLString(){
         $faker = new Generator();
         $faker->addProvider(new HtmlLorem($faker));
-        $node = $faker->randomHtml(6, 10);
+        $node = $faker->randomHtml(5, 5);
+        $dom = new \DOMDocument();
+        $error = $dom->loadHTML($node);
```
Comment (WillGibson) [inline test/Faker/Provider/HtmlLoremTest.php]: I think `$isValidHTML`  would make more sense than `$error` as a variable name here and below.

**Comment by WillGibson** [comment]
Are, I was about to start making a a very similar PR.  If you want any help getting this through the pip @Zhastreaus , give me a shout.

**Comment by Zhastreaus** [comment]
@WillGibson nope this is not my PR, I think it's @rudkjobing 's

**Comment by WillGibson** [comment]
@rudkjobing I'm going to fork from your fork and see if I can get the builds passing as I'd like to use this feature.

**Comment by rudkjobing** [comment]
I must be blind `Fatal error: Call to a member function randomNumber() on a non-object in /home/travis/build/fzaninotto/Faker/src/Faker/Provider/HtmlLorem.php on line 153`

File: test/Faker/Provider/HtmlLoremTest.php [hunk 3]
```diff
@@ -20,10 +20,27 @@ public function testProvider()
     public function testRandomHtmlReturnsValidHTMLString(){
         $faker = new Generator();
         $faker->addProvider(new HtmlLorem($faker));
-        $node = $faker->randomHtml(6, 10);
+        $node = $faker->randomHtml(5, 5);
         $dom = new \DOMDocument();
         $error = $dom->loadHTML($node);
```
Comment (WillGibson) [inline test/Faker/Provider/HtmlLoremTest.php]: Maybe `$isValidHtml` instead of `$error` here too?

**Comment by WillGibson** [comment]
Looking good now @rudkjobing. Can't see any build info here now. Are they passing now you fixed the fatal error?

**Comment by WillGibson** [comment]
@rudkjobing I have a passing build here https://travis-ci.org/BiffBangPow/Faker/jobs/244895238. I'll try to submit a PR to your branch your with the fix.

**Comment by rudkjobing** [comment]
Awesome, thanks!

**Comment by WillGibson** [comment]
https://github.com/rudkjobing/Faker/pull/1

**Comment by rudkjobing** [comment]
Travis doesn't seem to run automatically?

**Comment by WillGibson** [comment]
It should run automatically because this is a PR I think, but you might need to log into Travis and add you repo/branch maybe?

**Comment by rudkjobing** [comment]
I havn't had issues with travis before. Don't really know what the problem is.

**Comment by WillGibson** [comment]
Me neither, and I can't play with your Travis builds directly I guess. It should build automatically under fzaninotto/Faker as it's a PR, at least that what I thought. @fzaninotto Can you help here? I had a passing build here (https://travis-ci.org/BiffBangPow/Faker/builds/244900976) if that's any use.

File: src/Faker/Provider/HtmlLorem.php [hunk 4]
```diff
@@ -66,6 +66,36 @@ public function randomHtml($maxDepth = 4, $maxWidth = 4)
         return $document->saveHTML();
     }
 
+    /**
+     * @param int $maxDepth
+     * @param int $maxWidth
+     *
+     * @return string
+     */
+    public function randomBody($maxDepth = 4, $maxWidth = 4)
```
Comment (WillGibson) [inline src/Faker/Provider/HtmlLorem.php]: Now that I'm using this in my actual developing code, I think this would be better called `randomHTMLBody`.

**Comment by rudkjobing** [comment]
Going to open a new pull request, to see if that helps.

**Comment by WillGibson** [comment]
Hi @rudkjobing 

Fingers crossed re the new PR.

Please forgive my amatuerish approach to commenting on PRs in GitHub, I'm just getting used to how it all works (we use BitBucket here at work so the UI is different) and trying to get a bit more involved, help out with open source stuff.

There's a couple more comments in there from me that I would like addressed unless you disagree with them, then I'm totally going to approve this PR which has saved me doing the same thing over here.

https://github.com/fzaninotto/Faker/pull/1159/files#diff-ecbf0f117b3c8b63635081589370cce6R91

https://github.com/fzaninotto/Faker/pull/1159/files#diff-b3f77ef5414bc4a1b3641d7567ae3998R25

Will

**Comment by rudkjobing** [comment]
Hi Will.

Don't worry! I did change the code based on your change requests. I made a new PR here: https://github.com/fzaninotto/Faker/pull/1237

**Pull request closed without merging**